eclair-core/src/main/scala/fr/acinq/eclair/channel/InteractiveTxBuilder.scala

```diff
@@ -0,0 +1,748 @@
+        } else {
+          filterInputs(fundedTx, currentInputs, unusableInputs)
+        }
+      case WalletFailure(t) =>
+        log.error("could not fund interactive tx: ", t)
+        // We use a generic exception and don't send the internal error to the peer.
+        replyTo ! LocalFailure(ChannelFundingError(fundingParams.channelId))
+        unlockAndStop(currentInputs.map(toOutPoint).toSet ++ unusableInputs.map(_.outpoint))
+      case msg: ReceiveMessage =>
+        timers.startSingleTimer(msg, 1 second)
+        Behaviors.same
+      case Abort =>
+        timers.startSingleTimer(Abort, 1 second)
+        Behaviors.same
+    }
+  }
+
+  /** Not all inputs are suitable for interactive tx construction. */
+  def filterInputs(fundedTx: Transaction, currentInputs: Seq[TxAddInput], unusableInputs: Set[UnusableInput]): Behavior[Command] = {
+    context.pipeToSelf(Future.sequence(fundedTx.txIn.map(txIn => getInputDetails(txIn, currentInputs)))) {
+      case Failure(t) => WalletFailure(t)
+      case Success(results) => InputDetails(results.collect { case Right(i) => i }, results.collect { case Left(i) => i }.toSet)
+    }
+    Behaviors.receiveMessagePartial {
+      case inputDetails: InputDetails =>
+        if (inputDetails.unusableInputs.isEmpty) {
+          // This funding iteration did not add any unusable inputs, so we can directly return the results.
+          val changeOutputs = fundedTx.txOut
+            .filter(_.publicKeyScript != fundingParams.fundingPubkeyScript)
+            .map(txOut => TxAddOutput(fundingParams.channelId, generateSerialId(), txOut.amount, txOut.publicKeyScript))
+          if (changeOutputs.length > 1) {
+            log.warn(s"bitcoind should never add more than one change output (${changeOutputs.length} found)")
+          }
+          val outputs = if (fundingParams.isInitiator) {
+            // If the initiator doesn't want to contribute, we should cancel out the dust amount artificially added previously.
+            val initiatorChangeOutputs = if (fundingParams.localAmount == 0.sat) {
+              changeOutputs.map(o => o.copy(amount = o.amount + fundingParams.dustLimit))
+            } else {
+              changeOutputs
+            }
+            // The initiator is responsible for adding the shared output.
+            TxAddOutput(fundingParams.channelId, generateSerialId(), fundingParams.fundingAmount, fundingParams.fundingPubkeyScript) +: initiatorChangeOutputs
+          } else {
+            // The protocol only requires the non-initiator to pay the fees for its inputs and outputs, discounting the
+            // common fields (shared output, version, nLockTime, etc). However, this is really hard to compute here,
+            // because we don't know the witness size of our inputs (we let bitcoind handle that). For simplicity's sake,
+            // we simply accept that we'll slightly overpay the fee (which speeds up channel confirmation).
+            changeOutputs
```

Comment:
> IIUC an implicit assumption is that the non-initiator has a non-zero contribution. 

Somewhat, yes. If we're in this method, it means we are currently adding funding to the transaction, which means we have a contribution as a non-initiator. 

> But we are still paying for it

I'm not sure what you mean here, we're not paying the fee for the shared output in this branch, only for our own change output. What this comment highlights is that as a non-initiator, with the current code, we're paying the fee for a **complete** transaction containing our own inputs and our change output (**without** the initiator's inputs, change output and the funding output), whereas according to the spec we could discount the common fields of the transaction (`nLockTime`, input/output count, etc) and pay slightly less fees, but we choose not to because it's tedious and should only marginally lower the fees we're paying as a non-initiator, so the complexity isn't worth the effort. Does that make more sense? Is there a re-wording of that comment that would be easier to understand?

> They all use native segwit, don't they?

But that doesn't give us their witness size though? It can very a lot for `pay2wsh` inputs, depending on what script they're using.

Reply:
> we're not paying the fee for the shared output in this branch, only for our own change output

But we did add `TxOut(toFund, fundingParams.fundingPubkeyScript)` to the `dummyTx` right? So we paid for it?

> we're paying the fee for a complete transaction [...] whereas according to the spec we could discount the common fields of the transaction (nLockTime, input/output count, etc) and pay slightly less fees

I understand better. I think it's the mention of "shared output" output as part of the "common fields" in the existing comment, which confused me. To summarize:
- common fields (`version`, `nLockTime`, input/output count, etc): totally fine with not discounting them because it's only a few bytes
- shared output: I think that with the current code the non-initiator pays for it, you are saying it doesn't.

Reply:
You are totally right on both points, thanks for highlighting this! I improved that in https://github.com/ACINQ/eclair/pull/2273/commits/370822c3da265f731f64efbe09f0a4823aa177de

We can indeed discount the extra fees in the case where we have a change output. We could also do it in the case where we don't have a change output, but it's really cumbersome because we would need to make an async call to get a change address, which is probably not worth the effort...